server/src/main/java/org/elasticsearch/reservedstate/service/FileSettingsService.java

```diff
@@ -0,0 +1,330 @@
+
+                            if (watchedFileChanged(path)) {
+                                processFileSettings(path, false);
+                            }
+                        } catch (Exception e) {
+                            logger.warn("error processing operator settings file", e);
+                        }
+                    } else {
+                        key.pollEvents();
+                        key.reset();
+                    }
+                }
+            } catch (ClosedWatchServiceException | InterruptedException expected) {
+                logger.info("shutting down watcher thread");
+            } catch (Exception e) {
+                logger.error("shutting down watcher thread with exception", e);
+            } finally {
+                watcherThreadLatch.countDown();
+            }
+        }, "elasticsearch[file-settings-watcher]").start();
+    }
+
+    synchronized void stopWatcher() {
+        logger.debug("stopping watcher ...");
+        if (watching()) {
+            try {
+                if (settingsDirWatchKey != null) {
+                    settingsDirWatchKey.cancel();
+                    settingsDirWatchKey = null;
+                }
+                fileUpdateState = null;
+                watchService.close();
+                if (watcherThreadLatch != null) {
+                    watcherThreadLatch.await();
+                }
+            } catch (IOException | InterruptedException e) {
+                logger.warn("encountered exception while closing watch service", e);
+            } finally {
+                watchService = null;
+                logger.info("watcher service stopped");
+            }
+        } else {
+            logger.debug("file settings service already stopped");
+        }
+    }
+
+    private WatchKey enableSettingsWatcher(WatchKey previousKey, Path settingsDir) throws IOException {
+        if (previousKey != null) {
+            previousKey.cancel();
+        }
+        return settingsDir.register(
+            watchService,
+            StandardWatchEventKinds.ENTRY_MODIFY,
+            StandardWatchEventKinds.ENTRY_CREATE,
+            StandardWatchEventKinds.ENTRY_DELETE
+        );
+    }
+
+    void processFileSettings(Path path, boolean onStartup) throws IOException {
+        logger.info("processing path [{}] for [{}]", path, NAMESPACE);
+        try (var json = new BufferedInputStream(Files.newInputStream(path))) {
+            try (XContentParser parser = JSON.xContent().createParser(XContentParserConfiguration.EMPTY, json)) {
+                stateService.process(NAMESPACE, parser, (e) -> {
+                    if (e != null) {
+                        // If we encountered an exception trying to apply the operator state at
+                        // startup time, we throw an error to force Elasticsearch to exit.
+                        if (onStartup) {
```

Comment:
The logic here is a little difficult to reason about since it depends on this boolean (but the error here is then not possible outside of startup). Perhaps a more direct approach would be to pass a Consumer<Exception> so the two callers can decide how to handle it.